### PR TITLE
switch images to alpine

### PR DESCRIPTION
**Pull request by Zoey2936**
# License Agreement for Contributions
By submitting this pull request, I acknowledge and agree that my contributions will be included in Stirling-PDF and that they can be relicensed in the future under MPL 2.0 (Mozilla Public License Version 2.0) license. 

(This does not change the general open-source nature of Stirling-PDF, simply moving from one license to another license)

### CODE REVIEW

**Comment by Frooodle** [comment]
Wow this is awesome work I could never get alpine images working due to inexperienced docker skills 

Have you been able to test these and the functionality of python/libre office etc?

**Comment by Frooodle** [comment]
Just ran a quick run of all 3 docker files

Ultra lite seems to be missing java

**Comment by Frooodle** [comment]
Cool with that it seems they all load to the homepage
image

I will do some tests on OCR and conversions just to see if anything else wrong in parallel with your testing etc

**Comment by Zoey2936** [comment]
From my tests (of the full image) everything works, except one thing; If I want to convert a pdf to word, I see in the docker logs: `Error: source file could not be loaded`  and it downloads an empty zip file

**Comment by Frooodle** [comment]
doing docker compose i can expose the port 
but when running in docker desktop i get
image
Without the Expose 8080

What is the norm for docker files

**Comment by Zoey2936** [comment]
docker run --rm -it p 8080:8080 <image-name> works for me on docker desktop

**Comment by Frooodle** [comment]
(Meaning i am unable to run it and access 8080)

**Comment by Frooodle** [comment]
yeah works via CMD for me as well, just find it odd that docker desktop UI doesnt work with it or let you define ports, so was wondering if this is something non standard

**Comment by Zoey2936** [comment]
I've never used EXPOSE in a Dockerfile since it does nothing: https://docs.docker.com/engine/reference/builder/#expose

**Comment by Zoey2936** [comment]
I've tried again, it downloads now the converted document, but it is unuseable (pdf => html:
grafik

**Comment by Frooodle** [comment]
Something is defo off with the outputs when compared to old version
(You can test here)
https://pdf.adminforge.de/pdf-to-word
pdf to work on your version gives me
image
vs
image


seems its just loaded the PDF file bytes into .doc format not actual converting

**Comment by Frooodle** [comment]
i see libreoffice-core is not avaible, i tried just 'libreoffice'
and get a slightly better conversion now.. all 3 pages are overlayed onto single page and rotation is not kept.. so something weird..
but clearly something different about the libre package

image

**Comment by Zoey2936** [comment]
I've found the error

File: Dockerfile-lite
Comment (Zoey2936) [inline Dockerfile-lite]: I've added WeasyPrint since it seemed to be small, no idea what it does / if it is needed / adds more features

File: DockerfileBase
Comment (Zoey2936) [inline DockerfileBase:21]: what does opencv do?

File: DockerfileBase
Comment (Zoey2936) [inline DockerfileBase]: are `uno unoconv WeasyPrint` needed since it seems to also work without it

Comment (Frooodle) [reply DockerfileBase:23]: uno and unoconv are not used now, we can remove!
WeasyPrint  is only for html to pdf

Comment (Frooodle) [reply DockerfileBase]: Sorry i take back what i said

Comment (Frooodle) [reply DockerfileBase]: x file to PDF uses uno and unoconv

Comment (Zoey2936) [reply DockerfileBase]: reverted

**Comment by Frooodle** [comment]
So right now the conversions are not happening in a way which i am happy with, it seems to throw everything onto single page, not sure why... Until then we wont be able to merge this.
However super appreciate your work, will try see whats up with libreoffice

**Comment by Zoey2936** [comment]
it seems to be a bug in libreoffice, so it will also affect debian/ubuntu when they will update in the future

**Comment by Frooodle** [comment]
If it's the version and not general package can we pull the older version?

**Comment by Zoey2936** [comment]
v7.3.7.2-r0 and earlier of libreoffice works, everything after v7.5.5.2-r0 not, so the bug seems to be added with v7.4 or v7.5

**Comment by Zoey2936** [comment]
https://packages.ubuntu.com/jammy/libreoffice currently the dokcer image uses v7.3

**Comment by Zoey2936** [comment]
using alpine:3.17 instead of :latest makes it work for lite and normal, ultra lite can stay at latest, but 3.17 is two versions old...

**Comment by Zoey2936** [comment]
https://mirror1.hs-esslingen.de/pub/Mirrors/tdf/libreoffice/src/bugs-changelog-tag-libreoffice-7.6.4.1-release-7.6.4.1.log
"PDF: Conversion of pdf to docx or doc collapses all content onto one page (tdf#157589) [Kevin Suo]"

**Comment by Zoey2936** [comment]
https://gitlab.alpinelinux.org/alpine/aports/-/issues/15628

File: Dockerfile-ultra-lite
Comment (Zoey2936) [inline Dockerfile-ultra-lite]: I preffer pinning version, but it seems this repo isn't manged by dependabot/renovate, etc. right?

File: Dockerfile-ultra-lite
Comment (Zoey2936) [inline Dockerfile-ultra-lite:23]: is there a reason why this is disabled? Can it be removed? Or should I try to get it working in PR after this is merged?

File: Dockerfile-lite
Comment (Zoey2936) [reply Dockerfile-lite]: maybe it should be mentioned in the versions guide

File: Dockerfile-ultra-lite
Comment (Frooodle) [reply Dockerfile-ultra-lite:23]: Previously i had some issues with it and not had time to resolve
I would love to have correct permissions and user applied

File: Dockerfile-ultra-lite
Comment (Frooodle) [reply Dockerfile-ultra-lite]: dependabot is running, but i believe its only reviewing the java gradle versions not docker, let me see if i can get it for both

File: DockerfileBase
Comment (Frooodle) [reply DockerfileBase:21]: opencv is for 
"detect blank pages"
and "Split photos"

File: Dockerfile-ultra-lite
Comment (Frooodle) [reply Dockerfile-ultra-lite]: Added for docker https://github.com/Stirling-Tools/Stirling-PDF/commit/ee96d2a0e3b79c5bd577c5b6f33d329cb42522a5

Comment (Frooodle) [reply Dockerfile-ultra-lite]: Can you pin version, would be best

Comment (Zoey2936) [reply Dockerfile-ultra-lite]: are you sure that dependabot also detect files which are not named Dockerfile? I only know it from renovate, renovate only dedects Dockerfiles when theier filename ends with Dockerfile (so it dedects Dockerfile, abc.Dockerfile, but not Dockerfile-abc), not sure how dependabot handels this

Comment (Frooodle) [reply Dockerfile-ultra-lite]: https://github.com/Stirling-Tools/Stirling-PDF/pull/635/files
https://github.com/Stirling-Tools/Stirling-PDF/pull/636/files
seems so

File: Dockerfile-lite
Comment (Frooodle) [reply Dockerfile-lite]: Weasyprint is for html to pdf

Comment (Zoey2936) [reply Dockerfile-lite]: should I keep it?

**Comment by Frooodle** [comment]
> using alpine:3.17 instead of :latest makes it work for lite and normal, ultra lite can stay at latest, but 3.17 is two versions old...

Based on this and you using 3.19 is this PR now blocked on resolution of alpine updating its package manager?

File: Dockerfile-lite
Comment (Frooodle) [reply Dockerfile-lite]: Yes i dont want to lose any functionality

**Comment by Zoey2936** [comment]
> > using alpine:3.17 instead of :latest makes it work for lite and normal, ultra lite can stay at latest, but 3.17 is two versions old...
> 
> Based on this and you using 3.19 is this PR now blocked on resolution of alpine updating its package manager?

yes blocked until they fix libreoffice, but from what I know, they are fast in fixing this

File: Dockerfile-lite
Comment (Zoey2936) [reply Dockerfile-lite]: I ask because lite didn't included Weasyprint before

**Comment by Zoey2936** [comment]
should now work, I will create a PR when a fixed libreoffice verion from outside edge can be used

**Comment by Frooodle** [comment]
btw wondering if i can get  your advice,
I am trying to get calibre installed on alpine

I get
```shell
apk add --no-cache calibre --repository=http://dl-cdn.alpinelinux.org/alpine/edge/testing
fetch http://dl-cdn.alpinelinux.org/alpine/edge/testing/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.18/main/x86_64/APKINDEX.tar.gz
fetch https://dl-cdn.alpinelinux.org/alpine/v3.18/community/x86_64/APKINDEX.tar.gz
ERROR: unable to select packages:
  py3-pyqt6-webengine (no such package):
    required by: calibre-7.3.0-r0[py3-pyqt6-webengine]
  qt6-qtwebengine (no such package):
    required by: calibre-7.3.0-r0[qt6-qtwebengine]
  so:libicui18n.so.74 (no such package):
    required by: calibre-7.3.0-r0[so:libicui18n.so.74]
  so:libicuuc.so.74 (no such package):
    required by: calibre-7.3.0-r0[so:libicuuc.so.74]
  so:libpodofo.so.2 (no such package):
    required by: calibre-7.3.0-r0[so:libpodofo.so.2]
 ```
    

As a solution i added

```
http://dl-cdn.alpinelinux.org/alpine/edge/main
http://dl-cdn.alpinelinux.org/alpine/edge/community
``` 
to
``/etc/apk/repositories``

is this okay?

**Comment by Zoey2936** [comment]
I would do this:
/etc/apk/repositories:
```
https://dl-cdn.alpinelinux.org/alpine/v3.19/main
https://dl-cdn.alpinelinux.org/alpine/v3.19/community

@testing https://dl-cdn.alpinelinux.org/alpine/edge/main
@testing https://dl-cdn.alpinelinux.org/alpine/edge/community
@testing https://dl-cdn.alpinelinux.org/alpine/edge/testing
```
`apk add --no-cache calibre@testing`

**Comment by Zoey2936** [comment]
can I ask when this PR will be merged?-

File: Dockerfile-lite
Comment (Zoey2936) [inline Dockerfile-lite]: what does python include?

Comment (Zoey2936) [reply Dockerfile-lite]: since I've added WeasyPrint

Comment (Frooodle) [reply Dockerfile-lite]: python includes cv2 stuff for
https://github.com/Stirling-Tools/Stirling-PDF/blob/main/scripts/split_photos.py
https://github.com/Stirling-Tools/Stirling-PDF/blob/main/scripts/detect-blank-pages.py

**Comment by Frooodle** [comment]
> can I ask when this PR will be merged?-

Once i figure out calibre and wkhtmltopdf
I added them as custom installs which users can do post docker image install
https://github.com/Stirling-Tools/Stirling-PDF/pull/682/files#diff-8e6e46e251822ef45509ec061b133c5788a5719a0670ac3158b49dbb58e3dab6R69

**Comment by Zoey2936** [comment]
should I add them to this PR? calibre and wkhtmltopdf?

**Comment by Frooodle** [comment]
If you can make it conditional in docker file via some arg in the init scripts or changed to match in the java file i linked then sure!

**Comment by Zoey2936** [comment]
should I add them to this PR? calibre and wkhtmltopdf?

> If you can make it conditional in docker file via some arg in the init scripts or changed to match in the java file i linked then sure!

waht do you mean?

File: Dockerfile-lite
Comment (Zoey2936) [reply Dockerfile-lite]:
```suggestion
ENV ENDPOINTS_GROUPS_TO_REMOVE=OpenCV,OCRmyPDF
```
since WeasyPrint is included

**Comment by Zoey2936** [comment]
what does wkhtmltopdf do? Is it not the same as WeasyPrint? And I thought wkhtmltopdf is deprecated

**Comment by Frooodle** [comment]
Sorry missed comments

WeasyPrint does a pretty bad job, i have had requests to use wkhtmltopdf as it produces vastly better pdf files, i have it as an optional download due to its high security issues
so its not recommended other than for local users.. i guess a better solution would be chromium or puppeteer though....?

"waht do you mean? should I add calibre to the full image?"

So for this i mean that they are optional based on user parameters, i woudlnt want them as part of image but downloaded as a optional step post image via init script or the app itself

**Comment by Zoey2936** [comment]
wkhtmltopdf is removed in newer alpine versions... - but I could add calibre to be installed if an env is set

**Comment by Frooodle** [comment]
> wkhtmltopdf is removed in newer alpine versions... - but I calibre can be added by running `apk add --no-cache calibre@testing`
> I've looked into the https://github.com/Stirling-Tools/Stirling-PDF/blob/main/src/main/java/stirling/software/SPDF/config/PostStartupProcesses.java - I think it would need to be changed completely to work with alpine - and I want to ask why you set the timezone in it?

I had issues with calibre install requiring user prompts without it

**Comment by Zoey2936** [comment]
what do you mean? maybe try adding `--no-interactive`

**Comment by Frooodle** [comment]
> what do you mean? maybe try adding `--no-interactive`

that didnt work for me, at least via ubuntu apt

**Comment by Zoey2936** [comment]
I talk about apk

**Comment by Frooodle** [comment]
Okay lets remove wkhtmltopdf  completly thats fine, ill merge your changes today or tomorrow and remove wkhtmltopdf as well as add
apk add --no-cache calibre@testing

**Comment by Frooodle** [comment]
was just rebuilding the dockerbaseFile (I think we can remove this base file now doesnt serve much purpose)

I got
2.066 fetch https://dl-cdn.alpinelinux.org/alpine/v3.19/community/x86_64/APKINDEX.tar.gz
2.242 ERROR: unable to select packages:
2.245   so:libpoppler.so.133 (no such package):
2.245     required by: libreoffice-draw-7.6.4.1-r0[so:libpoppler.so.133]

**Comment by Zoey2936** [comment]
please retry

**Review by Frooodle — APPROVED** [review]

**Comment by Frooodle** [comment]
https://github.com/Stirling-Tools/Stirling-PDF/actions/runs/7950256132/job/21702556824

gcc -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DFFI_BUILDING=1 -I/usr/include/ffi -I/usr/include/libffi -I/usr/include/python3.11 -c src/c/_cffi_backend.c -o build/temp.linux-aarch64-cpython-311/src/c/_cffi_backend.o
#22 98.48       error: command 'gcc' failed: No such file or directory
#22 98.48       [end of output]
#22 98.48   
#22 98.48   note: This error originates from a subprocess, and is likely not a problem with pip.
#22 98.48   ERROR: Failed building wheel for cffi
#22 98.48 Failed to build cffi
#22 98.48 ERROR: Could not build wheels for cffi, which is required to install pyproject.toml-based projects
#22 ERROR: process "/bin/sh -c echo \"@testing [https://dl-cdn.alpinelinux.org/alpine/edge/main\](https://dl-cdn.alpinelinux.org/alpine/edge/main/)" | tee -a /etc/apk/repositories &&     echo \"@testing [https://dl-cdn.alpinelinux.org/alpine/edge/community\](https://dl-cdn.alpinelinux.org/alpine/edge/community/)" | tee -a /etc/apk/repositories &&     echo \"@testing [https://dl-cdn.alpinelinux.org/alpine/edge/testing\](https://dl-cdn.alpinelinux.org/alpine/edge/testing/)" | tee -a /etc/apk/repositories &&     apk add --no-cache         ca-certificates         tzdata         tini         bash         curl         openjdk17-jre    libreoffice@testing         python3 &&         wget https://bootstrap.pypa.io/get-pip.py -qO - | python3 - --break-system-packages --no-cache-dir --upgrade &&     pip install --break-system-packages --no-cache-dir --upgrade unoconv WeasyPrint &&     mkdir -p /configs /logs /customFiles /pipeline/watchedFolders /pipeline/finishedFolders &&     fc-cache -f -v &&     chmod +x /scripts/*.sh" did not complete successfully: exit code: 1


Any idea? works for other images

**Comment by Frooodle** [comment]
just not lite

**Comment by Zoey2936** [comment]
try to add `build-base` at the beginning using apk and remove it at the end using `apk del --no-cache build-base`
and please do everything inside one RUN step to make the image smaller

**Comment by Frooodle** [comment]
@Zoey2936 for running as non root user how would be best with volume mounting etc?

I got stirling PDF running as none root user fine by using gosu/su-exec  to switch from root to stirlingpdf user via entrypoint script 
But cant get it working by doing standalone user

My concern is that i volume mount folders such as /logs which the app (started by stirling user) doesnt have permission to write to as since its volume mounted it users the hosts permissions and often uses root.

I can fix this by changing the permissions on host side but that would mess everyones installs and i wouldnt want to force people to change permissions on host level for every install

So is running root then switch to new user the only way? (since i can use root user to set volume mounted folders to stirling user after startup)
Basically

docker file
```
ENV PUID=1000 \
    PGID=1000 \
    UMASK=022 
RUN apk add --no-cache su-exec && \
    addgroup -g $PGID stirlingpdfgroup && \
    adduser -u $PUID -G stirlingpdfgroup -s /bin/sh -D stirlingpdfuser && \
    mkdir -p $HOME && chown stirlingpdfuser:stirlingpdfgroup $HOME
ENTRYPOINT ..... etc
```

entrypoint.sh
```
echo "Setting permissions and ownership for necessary directories..."
chown -R stirlingpdfuser:stirlingpdfgroup /logs /scripts /usr/share/fonts/opentype/noto /usr/share/tessdata /configs /customFiles
chmod -R 755 /logs /scripts /usr/share/fonts/opentype/noto /usr/share/tessdata /configs /customFiles

# Run the main command and switch to stirling user for rest of run
exec su-exec stirlingpdfuser "$@"
```

**Comment by Frooodle** [comment]
@Zoey2936  Seems 3.19.1 has the libreoffice bug again regardless of repository used,
If i move docker to tag 20240329 it fixes this but now this one has a issue with a python install as part of libreoffice..

```
0.648                py3-cffi-pyc-1.16.0-r1[python3~3.12]
0.648                py3-lxml-5.1.0-r0[python3~3.12]
0.648                py3-charset-normalizer-pyc-3.3.2-r1[python3]
0.648                py3-charset-normalizer-pyc-3.3.2-r1[python3~3.12]
0.649                py3-pygments-pyc-2.17.2-r1[python3]
0.649                py3-pygments-pyc-2.17.2-r1[python3~3.12]
0.650   so:libpython3.11.so.1.0 (no such package):
0.650     required by: libreoffice-common-7.6.4.1-r2[so:libpython3.11.so.1.0]
0.650   so:libmbedcrypto.so.7 (no such package):
0.650     required by: librist-0.2.10-r0[so:libmbedcrypto.so.7]
```

Any ideas?

**Comment by Zoey2936** [comment]
just wait... they updated to python 3.12 and their builder currently rebuilds all packages against python 3.12

**Comment by Frooodle** [comment]
Yeah first part seems fixed now thanks for bringing that to attention
And now libreoffice isnt including soffice and other executables :(

**Comment by Zoey2936** [comment]
What di you mean?

**Comment by Zoey2936** [comment]
So it doesn't work?

**Comment by Zoey2936** [comment]
I don't see any breaking chages in the last weeks https://gitlab.alpinelinux.org/alpine/aports/-/commits/master/community/libreoffice?ref_type=heads

**Comment by Frooodle** [comment]
How odd..
basically what i see is that after adding 
``` 
openssl \
openssl-dev \
    	```
    	
To the APK installs in current docker file as it seems to require that, when I check ``usr/lib/libreoffice/program/``
I would normally see soffice and several others and these are suddenly missing

**Comment by Zoey2936** [comment]
https://pkgs.alpinelinux.org/contents?file=soffice&path=&name=&branch=edge but they exist

**Comment by Frooodle** [comment]
Then i don't understand why it doesn't show :(

**Comment by Frooodle** [comment]
Correct and the file itself isn't in the final docker build for some reason

**Comment by Zoey2936** [comment]
there is also an issue opened: https://gitlab.alpinelinux.org/alpine/aports/-/issues/16005

**Pull request merged**
